### PR TITLE
Model Outputs - GV for Subtrim, Min, Max values #683

### PR DESCRIPTION
**Pull request by eshifri**
Implements #683

### CODE REVIEW

**Comment by pfeerick** [comment]
I broke it! 😆 

The scaling is wrong when you switch from GV back to "normal" - i.e. GV1 of 50 for channel min becomes 5.0. Plus (and more importantly!) it does not look like the GV values are actually being used (i.e. set output to use them and then check the channel monitor). I used GV1, GV2 and GV3, for subtrim, min and max values, set some default values and also used T5/T6 to adjust them up and down, and channel output was still what it had been set to numerically.

**Comment by eshifri** [comment]
Thank you!
I think it is only  a scaling factor: if I set GV1 to 90 chanel monitor goes to 90.
If I go back to number, output offset shows 9.0 and the channel monitor goes to 9.0
It is definitely not what you describe, but I do not see what we are doing differently.

**Comment by pfeerick** [comment]
Maybe this will help... `simu` is behaving exactly the same as hardware. 
https://user-images.githubusercontent.com/5500713/132438214-0fadb3fb-5b83-4a4f-9eb1-e1c91e46392e.mp4

**Comment by pfeerick** [comment]
I've verified against OTX Companion simulator, and whilst the radio GUI channel monitor breaks, altering the GV does alter channel min and max, so now need to see if the same is happening with ETX, as if it is, it's a UI bug this PR has exposed ;) Subtrim GV adjustment is working as expected though... I just hadn't tested that yet. 

Do you think the subtrim should be shown before the min and max on the overview screen? As it's currently in MIN, MAX, SUBTRIM order on the overview, and SUBTRIM, MIN, MAX when you edit it...

**Comment by eshifri** [comment]
I do not have strong opinion about the order of the fields; the windows look different enough to avoid confusion.

**Comment by pfeerick** [comment]
Right, have just tested the latest build for this again. Subtrim via GV is working perfectly. "Offset" when switching between GV and number is gone, which is good :) MIN and MAX values are still not working when set via GV - i.e. -50 for GV2, 50 for GV3 is not being applied. If switched back to number edit it does though. 

@MervDale Are you game to test this one since you wanted GV for outputs? Is it just me doing something wrong 😇 , or is this indeed not behaving 😱?

**Comment by eshifri** [comment]
I have the fix for limits. Will commit later today.

**Comment by pfeerick** [comment]
This is working very nearly perfectly now! :)

As part of testing I'm using T5 and T6 to adjust the GVs I'm using for min and max, and changing the GV whilst at the channel output screen *without* moving the stick at all will not update where the end limit bars are drawn... for that channel (but will for subtrim). As soon as that channel's stick is bumped or moved in any way (i.e. ADC drift), the limit bars are drawn correct. It works fine if the stick is held over whilst the GV is being changed though. 

I'm guessing `checkEvents()` could also invalidate if the limits have changed, not just on channel value like it currently does?

File: radio/src/gui/colorlcd/channel_bar.h [hunk 1]
```diff
@@ -145,13 +146,29 @@ class OutputChannelBar : public ChannelBar
       dc->drawSolidVerticalLine(width() / 2, 0, height(), COLOR_THEME_SECONDARY1);
 
       // Draw output limits bars
+      int limit = (g_model.extendedLimits ? LIMIT_EXT_MAX : 1000);
```
Comment (raphaelcoeffic) [inline radio/src/gui/colorlcd/channel_bar.h:149]: No magic values 😇

File: radio/src/gui/colorlcd/channel_bar.h [hunk 2]
```diff
@@ -179,10 +196,29 @@ class OutputChannelBar : public ChannelBar
         value = newValue;
         invalidate();
       }
+      int limit = (g_model.extendedLimits ? LIMIT_EXT_MAX : 1000);
+      LimitData* lim = limitAddress(channel);
+      
+      if (GV_IS_GV_VALUE(lim->min - 1000, -limit, 0)) {
+        int ldMin =
+            GET_GVAR_PREC1(lim->min - 1000, -limit, 0, mixerCurrentFlightMode) +
```
Comment (raphaelcoeffic) [inline radio/src/gui/colorlcd/channel_bar.h:204]: Maybe it's just me, but I prefer to have the sign (`+`) at the beginning of the number it is adding (helps me read better).

**Comment by raphaelcoeffic** [comment]
@eshifri we have a `clang-format` file that helps formating the code. I'm one of those guys editing code with fixed font in 2 columns on a 13" notebook 😅, so that I get annoyed by very long lines. This format limits to 80 columns and "normalizes" the code. But so we don't have huge diffs, it's ok for me if you pass only substantial changes to `clang-format`. Depending on your editor, support is more or less automatic (even with Emacs/Vim). Thx!

**Comment by eshifri** [comment]
@raphaelcoeffic I'm afraid I do not understand. :-(
Magic numbers - you mean 1000? This is use all over the system for the limits (exactly the same line). If I create a #define for this particular file it will be more confusing.
You mean lim->min -1000 instead of lim->min - 1000?

I thought I have applied formatting to all new code. I will check it.

**Comment by raphaelcoeffic** [comment]
> @raphaelcoeffic I'm afraid I do not understand. :-(
> Magic numbers - you mean 1000? This is use all over the system for the limits (exactly the same line). If I create a #define for this particular file it will be more confusing.
> You mean lim->min -1000 instead of lim->min - 1000?
> 
> I thought I have applied formatting to all new code. I will check it.

I understand, but we've got to start somewhere ;-)

**Comment by eshifri** [comment]
Hope it looks better now. :-)

**Review by raphaelcoeffic — APPROVED** [review]
The PR is yours @pfeerick

**Comment by pfeerick** [comment]
I'm just waking so will review shortly, and hope to merge today.

Since you use VSCode also, it should use the repo .clangformat config automatically, so it's safe to use "Format Selection" on new blocks of code, or substantial edits to existing stuff. 

If it weren't for overlap with other PRs (hence would cause merge conflicts) I'd have pushed a full file reformat commit just before merging... so maybe another time :grin:

**Comment by pfeerick** [comment]
Actually, no, I'm thinking of another PR (re: overlap). 

This is working great now @eshifri ... thank you for fixing up the last cosmetics! The channel monitor is now updating perfectly when the end limits are dynamically changed, and the output is working correct too, so this ones done! :)

**Pull request merged**